core/src/avm2/globals/json.rs

```diff
@@ -0,0 +1,443 @@
+//! `JSON` impl
+
+use crate::avm2::activation::Activation;
+use crate::avm2::array::ArrayStorage;
+use crate::avm2::class::Class;
+use crate::avm2::globals::array::ArrayIter;
+use crate::avm2::method::{Method, NativeMethodImpl};
+use crate::avm2::names::{Namespace, QName};
+use crate::avm2::object::{ArrayObject, FunctionObject, Object, ObjectPtr, TObject};
+use crate::avm2::value::Value;
+use crate::avm2::Error;
+use crate::ecma_conversions::f64_to_wrapping_i32;
+use crate::string::AvmString;
+use gc_arena::{GcCell, MutationContext};
+use json::{
+    codegen::Generator as JsonGenerator, object::Object as JsonObject, parse as parse_json,
+    JsonValue,
+};
+use std::cmp;
+use std::ops::Deref;
+
+fn deserialize_json_inner<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    json: JsonValue,
+    reviver: Option<Object<'gc>>,
+) -> Result<Value<'gc>, Error> {
+    Ok(match json {
+        JsonValue::Null => Value::Null,
+        JsonValue::Short(s) => {
+            AvmString::new_utf8(activation.context.gc_context, s.to_string()).into()
+        }
+        JsonValue::String(s) => AvmString::new_utf8(activation.context.gc_context, s).into(),
+        JsonValue::Boolean(b) => b.into(),
+        JsonValue::Number(num) => {
+            let num: f64 = num.into();
+            if num.fract() == 0.0 {
+                f64_to_wrapping_i32(num).into()
+            } else {
+                num.into()
+            }
+        }
+        JsonValue::Object(js_obj) => {
+            let obj_class = activation.avm2().classes().object;
+            let mut obj = obj_class.construct(activation, &[])?;
+            for entry in js_obj.iter() {
+                let key = AvmString::new_utf8(activation.context.gc_context, entry.0);
+                let val = deserialize_json_inner(activation, entry.1.clone(), reviver)?;
+                let mapped_val = reviver.map_or(Ok(val.clone()), |reviver| {
+                    reviver.call(None, &[key.into(), val], activation)
+                })?;
+                if matches!(mapped_val, Value::Undefined) {
+                    obj.delete_property(activation, &QName::new(Namespace::public(), key).into())?;
+                } else {
+                    obj.set_property(
+                        obj,
+                        &QName::new(Namespace::public(), key).into(),
```

Comment:
FYI we have `QName::dynamic_name(key)` as a shorthand.
(I may refactor the naming in the future :D )